src/permutations.rs

```diff
@@ -77,14 +97,11 @@ where
                     return None;
                 }
                 *state = PermutationState::Buffered { k, min_n: k };
-                Some(vals[0..k].to_vec())
+                Some(Item::extract_start(vals, k, k - 1))
             }
             PermutationState::Buffered { ref k, min_n } => {
                 if vals.get_next() {
-                    let item = (0..*k - 1)
-                        .chain(once(*min_n))
-                        .map(|i| vals[i].clone())
-                        .collect();
+                    let item = Item::extract_start(vals, *k, *min_n);
```

Comment:
Having a separate variant `Buffered` prevents us from using `PoolIndex`. Maybe always constructing `indices` might allow `PoolIndex` here, simplifying implementation and improving maintenance.